core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java

```diff
@@ -200,7 +211,104 @@ public void renameTable(TableIdentifier from, TableIdentifier to) {
+
+    if (isNamespace(nsPath)) {
+      throw new AlreadyExistsException("Namespace '%s' already exists!", namespace);
+    }
+
+    try {
+      fs.mkdirs(nsPath);
+
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Create namespace failed: %s", namespace);
+    }
+  }
+
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
+    Path nsPath = namespace.isEmpty() ? new Path(warehouseLocation)
+        : new Path(warehouseLocation, SLASH.join(namespace.levels()));
+    if (!isNamespace(nsPath)) {
+      throw new NoSuchNamespaceException("Namespace does not exist: %s", namespace);
+    }
+
+    try {
+      return Stream.of(fs.listStatus(nsPath))
+        .map(FileStatus::getPath)
+        .filter(path -> isNamespace(path))
+        .map(path -> append(namespace, path.getName()))
+        .collect(Collectors.toList());
+    } catch (IOException ioe) {
+      throw new RuntimeIOException(ioe, "Failed to list namespace under: %s", namespace);
+    }
+  }
+
+  private Namespace append(Namespace ns, String name) {
+    String[] levels = Arrays.copyOfRange(ns.levels(), 0, ns.levels().length + 1);
+    levels[ns.levels().length] = name;
+    return Namespace.of(levels);
+  }
+
+  @Override
+  public boolean dropNamespace(Namespace namespace) {
+    Path nsPath = new Path(warehouseLocation, SLASH.join(namespace.levels()));
+
+    if (!isNamespace(nsPath) || namespace.isEmpty()) {
+      throw new NoSuchNamespaceException("Namespace does not exist: %s", namespace);
+    }
+
+    try {
+      return fs.delete(nsPath, true /* recursive */);
+
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Namespace delete failed: %s", namespace);
+    }
+  }
+
+  @Override
+  public boolean alterNamespace(Namespace namespace, NamespaceChange... changes) {
+    throw new UnsupportedOperationException(
+        "Cannot alter namespace " + namespace + " : alterNamespace is not supported");
   }
 
+  @Override
+  public Map<String, String> loadNamespaceMetadata(Namespace namespace) {
+    Path nsPath = new Path(warehouseLocation, SLASH.join(namespace.levels()));
+
+    if (!isNamespace(nsPath) || namespace.isEmpty()) {
+      throw new NoSuchNamespaceException("Namespace does not exist: %s", namespace);
+    }
+
+    return ImmutableMap.of("location", nsPath.toString());
+  }
+
+  private boolean isNamespace(Path path) {
+    Path metadataPath = new Path(path, "metadata");
+    try {
+      return fs.isDirectory(path) && !(fs.exists(metadataPath) && fs.isDirectory(metadataPath) &&
+          (fs.listStatus(metadataPath, TABLE_FILTER).length >= 1));
+
+    } catch (IOException ioe) {
+      throw new RuntimeIOException(ioe, "Failed to list namespace info: %s ", path);
+    }
+  }
+
+  @Override
+  public boolean existsNamespace(Namespace namespace) {
+    Path nsPath = new Path(warehouseLocation, SLASH.join(namespace.levels()));
+    return isNamespace(nsPath) && !namespace.isEmpty();
```

Comment:
Why must the namespace be non-empty?

Reply:
Because in Hive, the namespace is only one level, so namespace have not sub namespace.  the sentence like that `show subNamespace from namespace` we should return an empty list not throw an error.